core/google/cloud/iam.py

```diff
@@ -0,0 +1,204 @@
+# Copyright 2017 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Non-API-specific IAM policy definitions
+
+For allowed roles / permissions, see:
+https://cloud.google.com/iam/docs/understanding-roles
+"""
+
+# Generic IAM roles
+
+OWNER_ROLE = 'roles/owner'
+"""Generic role implying all rights to an object."""
+
+EDITOR_ROLE = 'roles/editor'
+"""Generic role implying rights to modify an object."""
+
+VIEWER_ROLE = 'roles/viewer'
+"""Generic role implying rights to access an object."""
+
+
+class Policy(object):
+    """IAM Policy
+
+    See:
+    https://cloud.google.com/iam/reference/rest/v1/Policy
+
+    :type etag: str
+    :param etag: ETag used to identify a unique of the policy
+
+    :type version: int
+    :param version: unique version of the policy
+    """
+    _OWNER_ROLES = (OWNER_ROLE,)
+    """Roles mapped onto our ``owners`` attribute."""
+
+    _EDITOR_ROLES = (EDITOR_ROLE,)
+    """Roles mapped onto our ``editors`` attribute."""
+
+    _VIEWER_ROLES = (VIEWER_ROLE,)
+    """Roles mapped onto our ``viewers`` attribute."""
+
+    def __init__(self, etag=None, version=None):
+        self.etag = etag
+        self.version = version
+        self.bindings = {}
+
+    @property
+    def owners(self):
+        """Legacy access to owner role."""
+        result = set()
+        for role in self._OWNER_ROLES:
+            for member in self.bindings.get(role, ()):
+                result.add(member)
+        return result
```

Comment:
Just for clarity, let me summarize the current implementation:

- When creating a `Policy` instance from the resource returned by the `getIAMPolicy` API, it conflates uses with either the "legacy" `roles/owner` or and the "real" `roles/pubsub.admin` into its `owners` set.   Likewise for editors and viewers.  The reason for this implementation is that early on (two years back) there were users who had some topics (maybe subscriptions) with the legacy roles set, but the newer ones were entirely equivalent (i.e., mapped onto the same permissions relevant for topics/subscriptioions).

- When sending the resource *back* to the `setIAMPolicy` API, the policy discards the legacy roles because,  as @elibixby points out, they are not allowed to be set for pubsub topics or subscriptions (they were not allowed to be set even then).

So, nobody has been relying on setting legacy roles since before we added IAM support to pubsub:  it was never possible at all.  Silently converting existing legacy roles to the acceptable equivalents seemed like a good tradeoff at the time.  The other choice would have been to let the `setIAMPolicy` blow up if the user didn't clean out / replace the legacy roles.